Proposal: add IntlDatePatternGenerator

deltragon opened the pull request:
ICU exposes the [DateTimePatternGenerator](https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/classicu_1_1DateTimePatternGenerator.html) class that allows generating a formatting pattern from a given "skeleton" for a given locale.
This allows more flexibility than the current `$format` argument of `IntlDateFormatter::format()`, which takes either a few pre-defined constants or a hard-coded format.
This functionality also has been [requested](https://bugs.php.net/bug.php?id=70377) in the past.

For example, if an application requires a format that will always use the long version of a year, but the short version of a month (eg. "MM/dd/YYYY" for "en_US", or "dd.MM.YYYY" for "de_DE"), one is out of luck.
`IntlDateFormatter::SHORT` will use the short year for "de_DE" ("dd.MM.YY"), and `IntlDateFormatter::MEDIUM` will use the long version of a month for "en_US" ("MMM dd, YYYY").

With `IntlDatePatternGenerator::getBestPattern()`, a skeleton can be provided to generate the appropriate pattern for a given locale.
In the use-case given above, the skeleton "YYYYMMdd" will generate the desired patterns for each locale, which can then be passed to `IntlDateFormatter::format()`.

Note about the implementation:
This had been previously implemented as a static method on `IntlDateFormatter`, creating a new `DateTimePatternGenerator` on every call.
It is now a separate class that has to be instantiated before use. This leaves us open to implementing other methods/options provided by the underlying `DateTimePatternGenerator`. I am not sure if there is any use in those, however, as `getBestPattern()` seems to be the only one with a real-world use-case to me.

(Note: I am aware that this needs to go to internals/an RFC first. I'm currently having issues with my email provider that prevent me from subscribing to/sending to the list. I will send this proposal once these are resolved).

deltragon commented on ext/intl/tests/dateformat_get_best_pattern.phpt:
```diff
@@ -0,0 +1,36 @@
+--SKIPIF--
+<?php
+if (!extension_loaded('intl'))	die('skip intl extension not enabled'); ?>
+<?php if (version_compare(INTL_ICU_VERSION, '50.1.2', '<')) die('skip for ICU < 50.1.2'); ?>
```
As an aside:
This line is included in almost all intl dateformat tests - there, it is used in this form:
```php
<?php if (version_compare(INTL_ICU_VERSION, '50.1.2') >=  0) die('skip for ICU < 50.1.2'); ?>
```
This, however, ends up skipping the test on my machine (`INTL_ICU_VERSION === '67.1'`). I assume `version_compare` is used incorrectly here? Should this be fixed for all other tests as well (and switched to what is the more readable form in my opinion)? Is this line still necessary (as PHP 7.4 requires ICU 50.1?)

kocsismate replied:
I think only the skip message is a bit deceptive: if fact, `die('skip for ICU < 50.1.2');` means that the test is `skip`ped because it's `for ICU < 50.1.2`. It also took quite some time for me to get accustomed to this.

If ICU 50.1 is required now then these tests are probably  not in use anymore 🤔

kocsismate replied:
And I don't think the `version_compare`  is necessary for this test.

deltragon replied:
Ah yes, that explains the message.
I still dont understand why these tests are not required for newer ICU versions - I would expect basic tests like
https://github.com/php/php-src/blob/d4530f807a78d50c2f5fd66114632480e7651d31/ext/intl/tests/dateformat_format.phpt#L5
to still be in use for newer versions/after PHP 7.4.
(For an incomplete list of affected tests, this is the commit that startes skipping a bunch of them for ICU >= 51.2 (was later changes to 50.1): 4840b0a749454335218f6131892031333d66142d)

nikic replied:
Tests for other ICU versions are under _variant2, _variant3 etc.

nikic replied:
> If ICU 50.1 is required now then these tests are probably not in use anymore thinking

I believe the reason why these tests still exist is that our requirement is ICU 50.1, while this checks for ICU 50.1.2.

Though maybe we can bump up the requirements slightly, it looks like RHEL 7 nowadays ships with ICU 50.2 (according to https://rpms.remirepo.net/rpmphp/zoom.php?rpm=icu). @remicollet Would a 50.2 requirement be safe?

MaxSem replied:
The same RHEL version comes with PHP 5: https://rpms.remirepo.net/rpmphp/zoom.php?rpm=php so is RHEL even supposed to dictate what PHP 8 should do?

kocsismate commented on ext/intl/dateformat/dateformat.stub.php:
```diff
@@ -118,6 +118,12 @@ public static function formatObject($datetime, $format = null, ?string $locale =
+     * @return string|false
+     * @alias datefmt_get_best_pattern
+     */
+    public static function getBestPattern(string $skeleton, ?string $locale = null) {}
```
New methods can have a return type declaration. Return types are not declared for existing methods only because of BC purposes. See https://wiki.php.net/rfc/internal_method_return_types

kocsismate commented on ext/intl/tests/dateformat_get_best_pattern.phpt:
```diff
@@ -0,0 +1,36 @@
+
+echo IntlDateFormatter::getBestPattern(""), "\n";
+
+echo IntlDateFormatter::getBestPattern(), "\n";
```
Catching exceptions and then echoing the message is preferred. And then you can even get rid of the `EXPECTF` in favor of `EXPECT`.

kocsismate reviewed (commented):
Personally, I'd looking forward to an `IntlDateTimePatternGenerator ` class, even though it would expose only a small subset of all the ICU capabilities. IMO the static method on `IntlDateFormatter` doesn't really fit there.

deltragon commented:
@kocsismate Thank you for the review!
I have switched to a `IntlDateTimePatternGenerator` class as suggested, and addressed your other comments as well.
I have put the new files in `ext/intl/dateformat/` as well, since they belong there thematically - or would you put them in a separate folder?
There are also some names of internal methods/macros I am not entirely sure about, is it okay to leave them as is or should I try to be more consistent there? "DateTimePatternGenerator" is a rather unwieldy name, unfortunately.

kocsismate commented on ext/intl/dateformat/datetimepatterngenerator_methods.cpp:
```diff
@@ -0,0 +1,153 @@
+	}
+
+	INTL_CHECK_LOCALE_LEN_OR_FAILURE(locale_len);
+	if (locale_len == 0) {
```
IMO it's unfortunate that `IntlDateFormatter` fetches the default locale for both `null` and `""`. :/ I would be happy if it would only happen for `null`, but maybe we should just follow the already paved path... :S

deltragon replied:
Would `""` throw an exception in that case? That would be somewhat unintuitive in my opinion - as far as I can tell, ICU will accept any string as a locale and fall back to using the default locale, I'm not sure if an empty string should be the exception here.
Unless we also throw on invalid locales - I am not sure if that is easily possible, however.

kocsismate replied:
Ok, this makes sense to me, so my comment should be ignored.

kocsismate commented on ext/intl/dateformat/datetimepatterngenerator_methods.cpp:
```diff
@@ -0,0 +1,153 @@
+	DTPATTERNGEN_METHOD_FETCH_OBJECT_NO_CHECK;
+
+	if (dtpgo->dtpg != NULL) {
+		intl_errors_set(DTPATTERNGEN_ERROR_P(dtpgo), U_ILLEGAL_ARGUMENT_ERROR, "datetimepatterngenerator_create: cannot call constructor twice", 0);
```
What's the reason of this? I see that the check is from `IntlDateFormatter`, but I don't know why these are necessary.

deltragon replied:
For `IntlDateFormatter`, there is this test:
https://github.com/php/php-src/blob/8d743d5281c29e9750e183804b7ba02e1ff82f0b/ext/intl/tests/bug62081.phpt#L12-L13
How is that (anti-)pattern handled in other cases?

kocsismate commented on ext/intl/dateformat/datetimepatterngenerator_methods.cpp:
```diff
@@ -0,0 +1,153 @@
+
+	intl_stringFromChar(skeleton_uncleaned, skeleton_str, skeleton_len, DTPATTERNGEN_ERROR_CODE_P(dtpgo));
+
+	INTL_METHOD_CHECK_STATUS(dtpgo, "datetimepatterngenerator_get_best_pattern: skeleton was not a valid UTF-8 string");
```
```suggestion
	INTL_METHOD_CHECK_STATUS(dtpgo, "Skeleton is not a valid UTF-8 string");
```

So the function name shouldn't be added (it's superfluous), the message should start with a capital character, and AFAIR we don't use past tense

kocsismate replied:
Also, I'm wondering if we can consider an invalid UTF-8 skeleton a programmatic error? As far as I see, we can, because the input should come from the programmer, not from the user. If my hypothesis is ok, then we could throw an exception instead of these warnings, like this:
`zend_argument_value_error(2, "must be a valid UTF-8 string")` (the `2` should be adapted in case of the method counterpart, since `$skeleton` is the 1st parameter there).

Last year, we migrated hundreds of warnings to exceptions, so it would be nice to follow this practice here as well, if possible.

deltragon replied:
I have adjusted the messages.
I would agree that this is most likely a programming error - I used a warning here for consistency with similar cases (eg. `IntlDateFormatter::setPattern()`), but I can switch to using an exception if preferred.

kocsismate commented on ext/intl/dateformat/datetimepatterngenerator_class.cpp:
```diff
@@ -0,0 +1,149 @@
+/* }}} */
+
+/* {{{ clone handler for IntlDateTimePatternGenerator */
+static zend_object *IntlDateTimePatternGenerator_object_clone(zend_object *object)
```
Cloning could be tested as well

deltragon replied:
I have added a test.

kocsismate commented on ext/intl/dateformat/datetimepatterngenerator_class.cpp:
```diff
@@ -0,0 +1,149 @@
+/* {{{ IntlDateTimePatternGenerator_object_dtor */
+static void IntlDateTimePatternGenerator_object_dtor(zend_object *object )
+{
+	zend_objects_destroy_object( object );
```
I think this is the default `dtor_obj` handler, so there is no need to override it (the same should apply for `IntlDateFormatter`

deltragon replied:
Thank you, I have removed it.

kocsismate commented on ext/intl/dateformat/datetimepatterngenerator_class.cpp:
```diff
@@ -0,0 +1,149 @@
+	intern = (IntlDateTimePatternGenerator_object*)ecalloc(1, sizeof(IntlDateTimePatternGenerator_object) + sizeof(zval) * (ce->default_properties_count - 1));
+
+	zend_object_std_init(&intern->zo, ce);
+    object_properties_init(&intern->zo, ce);
```
nit: indentation

deltragon replied:
Fixed.

kocsismate commented on ext/intl/dateformat/datepatterngenerator_class.cpp:
```diff
@@ -0,0 +1,141 @@
+}
+/* }}} */
+
+/* {{{ void IntlDatePatternGenerator_object_init(IntlDatePatternGenerator_object* to)
```
Please remove these comments (as well as any `{{{` and `}}}`, they only exists because of historical reasons.

deltragon replied:
Thank you - I removed them.

deltragon commented:
Rebased and removed the legacy procedural API.

nikic commented on ext/intl/dateformat/datepatterngenerator_class.cpp:
```diff
@@ -0,0 +1,131 @@
+	zend_objects_clone_members(&dtpgo_new->zo, &dtpgo_orig->zo);
+
+	if (dtpgo_orig->dtpg != NULL) {
+		DateTimePatternGenerator	*newDtpg;
+
+		newDtpg = dtpgo_orig->dtpg->clone();
```
```suggestion
		DateTimePatternGenerator *newDtpg = dtpgo_orig->dtpg->clone();
```

nikic replied:
Also above, please combine variable declaration and initialization.

nikic commented on ext/intl/dateformat/datepatterngenerator_class.cpp:
```diff
@@ -0,0 +1,131 @@
+
+static zend_object *IntlDatePatternGenerator_object_create(zend_class_entry *ce)
+{
+	IntlDatePatternGenerator_object*	intern;
+
+	intern = (IntlDatePatternGenerator_object*)ecalloc(1, sizeof(IntlDatePatternGenerator_object) + sizeof(zval) * (ce->default_properties_count - 1));
```
```suggestion
	IntlDatePatternGenerator_object *intern = zend_object_alloc(sizeof(IntlDatePatternGenerator_object), ce);
```

nikic commented on ext/intl/dateformat/datepatterngenerator_class.cpp:
```diff
@@ -0,0 +1,131 @@
+	object_properties_init(&intern->zo, ce);
+	IntlDatePatternGenerator_object_init(intern);
+
+
```
Redundant newline.

nikic commented on ext/intl/dateformat/datepatterngenerator_class.h:
```diff
@@ -0,0 +1,69 @@
+#ifndef DATEPATTERNGENERATOR_CLASS_H
+#define DATEPATTERNGENERATOR_CLASS_H
+
+//redefinition of inline in PHP headers causes problems, so include this before
```
I believe this should no longer be needed.

nikic commented on ext/intl/dateformat/datepatterngenerator_class.h:
```diff
@@ -0,0 +1,69 @@
+	intl_error  err;
+
+	// ICU break iterator
+	DateTimePatternGenerator*	dtpg;
```
```suggestion
	DateTimePatternGenerator *dtpg;
```
Avoid tabs for alignment -- and this doesn't actually align with anything...

nikic commented on ext/intl/dateformat/datepatterngenerator_methods.cpp:
```diff
@@ -0,0 +1,145 @@
+		intl_error_set(NULL, DTPATTERNGEN_ERROR_CODE(dtpgo),
+				"Error creating DateTimePatternGenerator",
+				0);
+		goto error;
```
This goto error seems pretty pointless, seeing how it goes to the next line ^^

nikic replied:
Though can you just return FAILURE here and unconditionally return SUCCESS below?

nikic commented on ext/intl/dateformat/datepatterngenerator_methods.cpp:
```diff
@@ -0,0 +1,145 @@
+{
+	zend_error_handling error_handling;
+
+	zend_replace_error_handling(EH_THROW, IntlException_ce_ptr, &error_handling);
```
Do we need this error handling replacement?

deltragon replied:
I copied this from `ext/intl/dateformat/dateformat_create.cpp`, other classes in intl that have both `__construct` and `create` also seem to handle it the same way, eg. `Collator`.

nikic replied:
Okay, let's keep it. ext/intl is pretty arcane.

nikic commented on ext/intl/dateformat/datepatterngenerator_methods.cpp:
```diff
@@ -0,0 +1,145 @@
+
+	INTL_METHOD_CHECK_STATUS(dtpgo, "Error getting cleaned skeleton");
+
+	{
```
Code block unnecessary. As usual, you can directly combine declaration and initialization.

nikic commented:
Looks like this needs a rebase. While at it, it would be great to add a mention of the new class in the UPGRADING file.

deltragon commented:
Rebased.

jrfnl commented:
Just checking - am I reading things correctly when I say that the two procedural functions mentioned in the RFC were not implemented in the final merged implementation ?
https://wiki.php.net/rfc/intldatetimepatterngenerator#proposal

cmb69 commented:
@jrfnl, ugh, apparently not. I wonder whether that has been explicitly discussed, or was just forgotten.

deltragon commented:
@jrfnl That was discussed and dropped during the vote, see: https://externals.io/message/114473#114658 and the messages following it.

jrfnl commented:
Thanks @cmb69 and @deltragon for your quick responses. Just wanted to verify as I'm working on adding detecting for the PHP 8.1 features/deprecations to [PHPCompatibility](https://github.com/PHPCompatibility/PHPCompatibility/issues/1299).
